### PR TITLE
Compute Token and Icode constants

### PR DESCRIPTION
**Pull request by rPraml**
This is an alternate approach for solving the problems mentioned here #1639

and should avoid large renumbering as you can see here: https://github.com/mozilla/rhino/pull/1593/files#diff-c1caeb9d13c1dceec5f4603f3836f3e227350ee1a57679238ce9a3a434601b60

### CODE REVIEW

File: rhino/src/main/java/org/mozilla/javascript/Token.java [hunk 1]
```diff
@@ -32,205 +32,206 @@ public static enum CommentType {
...
+            NULL = STRING + 1,
+            THIS = NULL + 1,
+            FALSE = THIS + 1,
+            TRUE = FALSE + 1,
+            SHEQ = TRUE + 1, // shallow equality (===)
+            SHNE = SHEQ + 1, // shallow inequality (!==)
+            REGEXP = SHNE + 1,
+            BINDNAME = REGEXP + 1,
+            THROW = BINDNAME + 1,
+            RETHROW = THROW + 1, // rethrow caught exception: catch (e if ) use it
```
Comment (rPraml) [inline rhino/src/main/java/org/mozilla/javascript/Token.java:89]: When inserting new opcodes, only the line before and after the insertion point has to be changed

**Comment by gbrail** [comment]
This is awfully clever -- I have not seen this before in Java and I'm wondering why not! I'll try it tonight.

**Comment by gbrail** [comment]
Does anyone else have a concern about this -- I'm having trouble seeing the downside and would like to merge it.

**Comment by p-bakker** [comment]
LGTM (we use the same approach in JS running in Rhino works like a charm 🙂)

**Comment by rbri** [comment]
> Does anyone else have a concern about this -- I'm having trouble seeing the downside and would like to merge it.

+1

We use this pattern since years for JDBC index based access of result sets (sometimes using the index is a bit faster)

**Comment by rPraml** [comment]
>  I'm having trouble seeing the downside and would like to merge it.

The only downside is IMHO, you must understand the concept and add new constants the correct way (which should be more straight forward as before)
The produced class file should be the same, as the compiler resolves the constants at compile time.

**Comment by p-bakker** [comment]
One slightly other approach that we also use and that is imho easier to understand and less likely to go wrong when adding new codes:
```
MIN_ICODE = 0,
Icode_DELNAME = MIN_ICODE--,

// Stack: ... value1 -> ... value1 value1
Icode_DUP = MIN_ICODE--,

....

// Call to GetTemplateLiteralCallSite
Icode_TEMPLATE_LITERAL_CALLSITE = MIN_ICODE--,
Icode_LITERAL_KEYS = MIN_ICODE--,
Icode_LITERAL_KEY_SET = MIN_ICODE--;
```

**Comment by rPraml** [comment]
> One slightly different approachwe also use and that is imho easier to understand and less likely to go wrong when adding new codes (at the cost of declaring another (private) variable):
> 
> ```
> private static int count = 0,
> 
> static final int
>     // delete operator used on a name
>     Icode_DELNAME = count--,
> ...
> ```
Unfortunately, this does not work here, as the result isn't computed at compile time, but on class instantiation. So these are no compile-time 'constants', that can be used in switch statements or annotations.

**Comment by p-bakker** [comment]
K, similar stuff in JavaScript dus do the trick, but Java ain't JavaScript or vise versa 🙂

**Comment by rbri** [comment]
@rPraml thanks a lot for this

**Pull request merged**